src/csvw/dsv.py

```diff
@@ -88,18 +115,40 @@ def __exit__(self, type, value, traceback):
         if self._close:
             self.f.close()
 
-    def writerow(self, row):
+    def writerow(self, row: typing.Union[tuple, list]):
         self.writer.writerow(self._escapedoubled(row))
 
-    def writerows(self, rows):
+    def writerows(self, rows: typing.Iterable[typing.Union[tuple, list]]):
         for row in rows:
             self.writerow(row)
 
 
 class UnicodeReader(object):
-    """Read Unicode data from a csv file."""
-
-    def __init__(self, f, dialect=None, **kw):
+    """
+    Read Unicode data from a csv file.
+
+    :param f: The source from which to read the data; a local path specified as `str` or \
+    `pathlib.Path`, a file-like object or a `list` of lines.
+    :param dialect: Either a dialect name as recognized by `csv.reader` or a \
+    :class:`~Dialect` instance for dialect customization beyond what can be done with \
+    `csv.writer`.
+    :param kw: Keyword arguments passed through to `csv.reader`.
+
+    .. code-block:: python
+
+        >>> with UnicodeReader('tests/fixtures/frictionless-data.csv', delimiter='|') as reader:
+        ...     for row in reader:
+        ...         print(row)
+        ...         break
+        ...
+        ['FK', 'Year', 'Location name', 'Value', 'binary', 'anyURI', 'email', 'boolean', 'array',
+        'geojson']
+    """
+    def __init__(
+            self,
+            f: typing.Union[str, pathlib.Path, typing.IO, typing.List[str]],
```

Comment:
Can we follow [stdlib](https://github.com/python/typeshed/blob/8e4b89a707f2508e4938e0691b714e235d1719fd/stdlib/_csv.pyi#L58) by accepting any iterable of lines (instead of a sequence, i.e. requiring the caller to materialize)?

nit: otherwise preferrably `typing.Sequence` (broad types for input arguments, concrete for return types, cf. Postel's law)

Reply:
Will be addressed with #66